src/DataProtection/DataProtection/src/KeyManagement/KeyRingBasedDataProtector.cs

```diff
@@ -381,6 +353,118 @@ public byte[] GetAadForKey(Guid keyId, bool isProtecting)
             }
         }
 
+#if NET10_0_OR_GREATER
+        public AdditionalAuthenticatedDataTemplate(string[] purposes)
+        {
+            // additionalAuthenticatedData := { magicHeader (32-bit) || keyId || purposeCount (32-bit) || (purpose)* }
+            // purpose := { utf8ByteCount (7-bit encoded) || utf8Text }
+
+            var keySize = sizeof(Guid);
+            int totalPurposeLen = 4 + keySize + 4;
+
+            var purposeLengthsPool = ArrayPool<int>.Shared.Rent(purposes.Length);
```

Comment:
Did you measure if it was worth renting an array only to prevent an extra call to `SecureUtf8Encoding.GetByteCount` and `Measure7BitEncodedUIntLength`

Reply:
You mean just pay the cost twice,? Yeah, should probably measure it.

Side thought: I'm guessing we don't actually expect many purposes. Another thing we could do here, if you're concerned about `ArrayPool` overhead, is use the stack - something like:

```
t[]? lease = null;
Span<int> lengths = purposes.Length <= 32 ? stackalloc int[purposes.Length] : (lease = ArrayPool<int>.Shared.Rent(purposes.Length)).AsSpan(0, purposes.Length);
...
if (lease is not null) ArrayPool<int>.Shared.Return(lease);
```

Reply:
[Benchmark code](https://github.com/DeagleGross/Baraholka/blob/main/Benchmarks/Benchmarks/DataConverterRewrite.cs)

| Method                                 | Mean     | Error   | StdDev  | Gen0   | Gen1   | Gen2   | Allocated |
|--------------------------------------- |---------:|--------:|--------:|-------:|-------:|-------:|----------:|
| MemoryStream                           | 141.4 ns | 2.62 ns | 3.31 ns | 0.0067 |      - |      - |     512 B |
| Manual                                 | 101.8 ns | 0.98 ns | 0.92 ns | 0.0023 | 0.0002 | 0.0002 |     320 B |
| Manual_CalculateByteCountEveryTime     | 124.7 ns | 1.62 ns | 1.52 ns | 0.0017 |      - |      - |     128 B |
| Manual_StackAllocForSmallPurposeArrays | 122.4 ns | 0.70 ns | 0.54 ns | 0.0014 |      - |      - |     128 B |

It seems like that last option Marc suggested is the most balanced - we indeed dont expect many purposes (I think), and this solution does not contest for the `ArrayPool` rent. 

What do you think?

Reply:
are the last two doing the same work? it seems odd that the perf is worse than `Manual` - we're doing *less* work (by not touching the pool at all) - why is it taking longer? question: can you try this with `[SkipLocalsInit(true)]` on the method? (this is applied globally in aspnetcore, I believe, but I don't know whether it'll apply to your benchmark which looks to be custom)

Reply:
> can you try this with [SkipLocalsInit] on the method? or the entire type for fairness (this is applied globally in aspnetcore, I believe

Don't think it is. https://github.com/dotnet/aspnetcore/issues/26586

Reply:
> > can you try this with [SkipLocalsInit] on the method? or the entire type for fairness (this is applied globally in aspnetcore, I believe
> 
> Don't think it is. #26586

We should revisit this, IMHO.